homeassistant/components/vesync/binary_sensor.py

```diff
@@ -0,0 +1,104 @@
+"""Binary Sensor for VeSync."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from pyvesync.vesyncbasedevice import VeSyncBaseDevice
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .common import rgetattr
+from .const import DOMAIN, VS_COORDINATOR, VS_DEVICES, VS_DISCOVERY
+from .coordinator import VeSyncDataCoordinator
+from .entity import VeSyncBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class VeSyncBinarySensorEntityDescription(BinarySensorEntityDescription):
+    """A class that describes custom binary sensor entities."""
+
+    is_on: Callable[[VeSyncBaseDevice], bool]
+
+
+SENSOR_DESCRIPTIONS: tuple[VeSyncBinarySensorEntityDescription, ...] = (
+    VeSyncBinarySensorEntityDescription(
+        key="water_lacks",
+        translation_key="water_lacks",
+        is_on=lambda device: device.water_lacks,
+        device_class=BinarySensorDeviceClass.PROBLEM,
+    ),
+    VeSyncBinarySensorEntityDescription(
+        key="details.water_tank_lifted",
+        translation_key="water_tank_lifted",
+        is_on=lambda device: device.details["water_tank_lifted"],
+        device_class=BinarySensorDeviceClass.PROBLEM,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up binary_sensor platform."""
+
+    coordinator = hass.data[DOMAIN][VS_COORDINATOR]
+
+    @callback
+    def discover(devices):
+        """Add new devices to platform."""
+        _setup_entities(devices, async_add_entities)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, VS_DISCOVERY.format(VS_DEVICES), discover)
+    )
+
+    _setup_entities(hass.data[DOMAIN][VS_DEVICES], async_add_entities, coordinator)
+
+
+@callback
+def _setup_entities(devices, async_add_entities, coordinator):
+    """Add entity."""
+    async_add_entities(
+        (
+            VeSyncBinarySensor(dev, description, coordinator)
+            for dev in devices
+            for description in SENSOR_DESCRIPTIONS
+            if rgetattr(dev, description.key) is not None
```

Comment:
I find the use of key as means to check attribute presence cryptic. Why not use device type based so that we generate sensors for known cases? It looks like a lot more attributes are populated in [build_humid_dict](https://github.com/webdjoe/pyvesync/blob/736c29256a5a8b562ddb1b16ac20e418b3e27f29/src/pyvesync/vesyncfan.py#L2523) with default value but I definite know that my humidifier does not support any of them.

Reply:
Thinking more on this - maybe I move that same logic to a exists_fn? Still with the reget but within the exists lambda.   That way it is flexible depending on what is needed.

Reply:
is_on is binary sensor native name: https://developers.home-assistant.io/docs/core/entity/binary-sensor

Now that I look at it again I think I leave as is. Not sure an easier way exists to use the is on value to safely check it?

Reply:
The new platform works correctly.

Reply:
To confirm do you have changes requested or are you okay with approving as is?

Reply:
I would loved to see key and value lambda be combined since they seem to be using the same data path but I am okay with what we have.

Reply:
Got it.  So combine in description and use the same value programmatically for "is_on". I will give it a go.

Reply:
It was just an idea. If we can have cases where key is one thing but the data returned is something, then we would want to keep what we have.

Reply:
It is possible that occurs but quick to make that change when/if it occurs. VScode is uploading the changes now.   Should be up shortly - it takes forever.